src/GafferSceneUI/ImageSelectionTool.cpp

```diff
@@ -0,0 +1,782 @@
+ImageGadget *ImageSelectionTool::imageGadget()
+{
+	return runTimeCast<ImageGadget>( view()->viewportGadget()->getPrimaryChild() );
+}
+
+
+void ImageSelectionTool::plugDirtied( const Gaffer::Plug *plug )
+{
+	if( plug == imagePlug()->metadataPlug() )
+	{
+		// We use this dirty signal in a somewhat unconventional way : we don't actually care about whether
+		// the metadata has changed, but we assume that if the input image is modified, it will trigger this
+		// signal. That gives us a cue that we need to recheck the manifest - even if none of the metadata
+		// has changed, if the image was rewritten, the manifest might have been rewritten as well.
+		m_manifestDirty = true;
+
+		selectedPathsChanged();
+		setStatus( "", false );
+	}
+}
+
+IECore::PathMatcher ImageSelectionTool::pathsForIDs( const std::vector<uint32_t> &ids, std::string &message )
+{
+	if( !ids.size() )
+	{
+		return IECore::PathMatcher();
+	}
+
+	updateRenderManifest( message );
+	if( !m_renderManifest )
+	{
+		return PathMatcher();
+	}
+
+	return m_renderManifest->pathsForIDs( ids );
+}
+
+void ImageSelectionTool::idsForPaths( const IECore::PathMatcher &paths, std::vector<uint32_t> &result, std::string &message )
+{
+	if( !paths.size() )
+	{
+		result.clear();
+		return;
+	}
+
+	updateRenderManifest( message );
+
+	if( !m_renderManifest )
+	{
+		result.clear();
+		return;
+	}
+
+	// TODO - would it be worth saving an allocation by passing in `result` here?
+	result = m_renderManifest->idsForPaths( paths );
+}
+
+uint32_t ImageSelectionTool::pixelID( const Imath::V2i &pixel )
+{
+	const auto [ wipeEnabled, wipePosition, wipeDirection ] = effectiveWipePlane( imageGadget() );
+
+	if( wipeEnabled && ( Imath::V2f( pixel ) + Imath::V2f( 0.5f ) - wipePosition ).dot( wipeDirection ) > 0 )
+	{
+		return 0;
+	}
+
+	Context::Scope scopedContext( view()->context() );
+	GafferImage::Sampler sampler( imagePlug(), idChannelName( imagePlug()->channelNames()->readable() ), Box2i( pixel, pixel + V2i( 1 ) ) );
+	float floatID = sampler.sample( pixel.x, pixel.y );
+	uint32_t id;
+	memcpy( &id, &floatID, 4 );
+	return id;
+}
+
+std::unordered_set<uint32_t> ImageSelectionTool::rectIDs( const Imath::Box2i &rect )
+{
+	Context::Scope scopedContext( view()->context() );
+	const Imath::Box2i validRect = BufferAlgo::intersection( rect, imagePlug()->dataWindow() );
+
+	std::unordered_set<uint32_t> result;
+	if( BufferAlgo::empty( validRect ) )
+	{
+		return result;
+	}
+
+	const auto [ wipeEnabled, wipePosition, wipeDirection ] = effectiveWipePlane( imageGadget() );
+
+
+	GafferImage::Sampler sampler( imagePlug(), idChannelName( imagePlug()->channelNames()->readable() ), validRect );
+
+	float prevValue = sampler.sample( validRect.min.x, validRect.min.y );
+	uint32_t prevID;
+	memcpy( &prevID, &prevValue, 4 );
+	if( prevID != 0 )
+	{
+		result.insert( prevID );
+	}
+
+	sampler.visitPixels(
+		validRect,
+		[&result, &prevValue, &wipeEnabled, &wipePosition, &wipeDirection] ( float value, int x, int y )
```

Comment:
C++17 doesn't allow lambda to capture structured bindings - GCC will let you do it, but Clang refuses to compile. One workaround seems to be to do an "init capture" instead : `&wipeEnabled = wipeEnabled`.

Reply:
C++20 allows it of course, but as we all know, it isn't 2020 yet in these parts.